### PR TITLE
Add support FrankerFaceZ badges.

### PR DESCRIPTION
**Pull request by Mm2PL**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description

Add support for FrankerFaceZ badges, partially solving #1658 (issue requested BTTV too).

# Changes
 1. Introduce badge provider,
 2. Refactor `MessageElementFlag`s to be `int64_t` instead of `int` as there wasn't space for a new flag there, 
 3. Add `MessageElementFlag::BadgeFfz`,
 4. Create showBadgesFfz (`/appearance/badges/ffz`) setting, default is true,
 5. Introduce checkbox for the `showBadgesFfz` setting,
 6. Add a MessageLayoutElement for ffz badges, to color them properly.

### CODE REVIEW

File: src/messages/MessageElement.hpp [hunk 1]
```diff
@@ -26,91 +26,97 @@ using ImagePtr = std::shared_ptr<Image>;
...
     // for links
-    LowercaseLink = (1 << 29),
-    OriginalLink = (1 << 30),
+    LowercaseLink = (1LL << 29),
+    OriginalLink = (1LL << 30),
 
     // ZeroWidthEmotes are emotes that are supposed to overlay over any pre-existing emotes
     // e.g. BTTV's SoSnowy during christmas season
-    ZeroWidthEmote = (1 << 31),
+    ZeroWidthEmote = (1LL << 31),
```
Comment (leon-richardt) [inline src/messages/MessageElement.hpp:120]: I suggest leaving a comment here stating that bit 32 is already used by a flag.
```suggestion

// (1LL << 32) is being used by `BadgeFfz`, see above

```

**Review by alazymeme — CHANGES_REQUESTED** [review]
As discussed, the default ffz badge should have a background (#755000) to match the default style. Otherwise the badge shows just fine.
Web Chat:
FFZ Badge Webchat

Chatterino:
FFZ Badge Chatterino

File: src/widgets/settingspages/GeneralPage.cpp [hunk 2]
```diff
@@ -620,6 +620,7 @@ void GeneralPage::initLayout(SettingsLayout &layout)
     layout.addCheckbox("Vanity (prime, bits, subgifter)",
                        getSettings()->showBadgesVanity);
     layout.addCheckbox("Chatterino", getSettings()->showBadgesChatterino);
+    layout.addCheckbox("FrankerFaceZ", getSettings()->showBadgesFfz);
```
Comment (alazymeme) [inline src/widgets/settingspages/GeneralPage.cpp:623]:
```suggestion
    layout.addCheckbox("FFZ Supporter", getSettings()->showBadgesFfz);
```

Comment (Mm2PL) [reply src/widgets/settingspages/GeneralPage.cpp:623]: FFZ badges include not only supporters, but bots and FFZ Developers.

Comment (alazymeme) [reply src/widgets/settingspages/GeneralPage.cpp:623]: Ahh that makes sense, I didn't realise. So should we then list which badges that includes in brackets? (like the authority, etc badges)

**Comment by Mm2PL** [comment]
> As discussed, the default ffz badge should have a background (#755000) to match the default style. Otherwise the badge shows just fine.
@alazymeme solved in f0231d0

File: src/widgets/settingspages/GeneralPage.cpp (hunk 2 above)
Comment (Mm2PL) [reply src/widgets/settingspages/GeneralPage.cpp]: Resolved in dca2c37

**Comment by Mm2PL** [comment]
@pajlada, could you please review this PR? 🙂

**Review by pajlada — COMMENTED** [review]
Looks good from a quick review, will test in weekend

**Review by alazymeme — APPROVED** [review]
Just based on the functionality alone, this works fine.

File: src/providers/ffz/FfzBadges.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+FfzBadges::FfzBadges()
+{
+}
+
+boost::optional<EmotePtr> FfzBadges::getBadge(const UserId &id)
+{
+    auto it = badgeMap.find(id.string);
+    if (it != badgeMap.end())
+    {
+        return emotes[it->second];
```
Comment (pajlada) [inline src/providers/ffz/FfzBadges.cpp:30]: use `this->`

File: src/providers/ffz/FfzBadges.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    this->loadFfzBadges();
+}
+
+FfzBadges::FfzBadges()
+{
+}
+
+boost::optional<EmotePtr> FfzBadges::getBadge(const UserId &id)
+{
+    auto it = badgeMap.find(id.string);
```
Comment (pajlada) [inline src/providers/ffz/FfzBadges.cpp:27]: use `this->`

File: src/providers/ffz/FfzBadges.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    if (it != badgeMap.end())
+    {
+        return emotes[it->second];
+    }
+    return boost::none;
+}
+boost::optional<QColor> FfzBadges::getBadgeColor(const UserId &id)
+{
+    auto badgeIt = badgeMap.find(id.string);
```
Comment (pajlada) [inline src/providers/ffz/FfzBadges.cpp:36]: use `this->`

File: src/providers/ffz/FfzBadges.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    if (badgeIt != badgeMap.end())
+    {
+        auto colorIt = colorMap.find(badgeIt->second);
```
Comment (pajlada) [inline src/providers/ffz/FfzBadges.cpp:39]: use `this->`

File: src/providers/ffz/FfzBadges.hpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    virtual void initialize(Settings &settings, Paths &paths) override;
+    FfzBadges();
+
+    boost::optional<EmotePtr> getBadge(const UserId &id);
+    boost::optional<QColor> getBadgeColor(const UserId &id);
+
+private:
+    void loadFfzBadges();
+    std::map<QString, int> badgeMap;
+    std::vector<EmotePtr> emotes;
```
Comment (pajlada) [inline src/providers/ffz/FfzBadges.hpp:28]: maybe rename to `badges`? I know the underlying type is an Emote, but in essence what's actually stored is an image used as a badge

**Review by pajlada — CHANGES_REQUESTED** [review]

**Comment by Mm2PL** [comment]
@pajlada, 🔔 suggestions done, merge conflicts resolved. It's ready to merge. 🙂

**Comment by alazymeme** [comment]
PAJJED Merge it

**Review by pajlada — APPROVED** [review]

**Pull request merged**